Feature/simple search

Arkitecht opened the pull request:
This adds a simple search feature (when enabled in config) to the main discussions page (which will search all discussion names and posts) and the discussion page (which will search all posts).

This is a simple search using a like search

Arkitecht commented:
Sure, here are some screenshots:
<img width="1247" alt="Screen Shot 2022-10-13 at 8 16 40 AM" src="https://user-images.githubusercontent.com/66117/195593972-08d3a44c-2a14-4163-9394-10cac3914413.png">
<img width="1227" alt="Screen Shot 2022-10-13 at 8 16 52 AM" src="https://user-images.githubusercontent.com/66117/195593973-5279404e-347d-464d-8df6-7bed51caacb5.png">
<img width="1252" alt="Screen Shot 2022-10-13 at 8 17 10 AM" src="https://user-images.githubusercontent.com/66117/195593974-970e5d55-b868-434e-a1f7-b637bae8e63f.png">
<img width="1258" alt="Screen Shot 2022-10-13 at 8 17 24 AM" src="https://user-images.githubusercontent.com/66117/195593975-100f6777-2c7f-4dd1-a35f-afb6ae97bd3d.png">

andreaselia commented on resources/views/components/search.blade.php:
```diff
@@ -0,0 +1,23 @@
+@if (\Firefly\Features::enabled('search'))
+    <form method="GET">
+        <div>
+            <label for="search" class="block text-lg font-medium text-gray-700 hidden">Quick search</label>
```
"block" and "hidden" on the same element? Do you mean to remove these and use "sr-only" or something?

Arkitecht replied:
yes, will update

andreaselia commented on resources/views/discussions/show.blade.php:
```diff
@@ -32,6 +32,8 @@
 </div>
 
 <div class="space-y-5">
+    <x-search :search="$search"></x-search>
```
Can we do this instead or am I wrong in assuming that this is a thing?
```suggestion
    <x-search :search="$search" />
```

Arkitecht replied:
yes, will update

andreaselia commented on resources/views/index.blade.php:
```diff
@@ -5,6 +5,8 @@
     {{ __('Discussions') }}
 </h2>
 
+<x-search :search="$search"></x-search>
```
Same as above here.

andreaselia commented on src/Models/Discussion.php:
```diff
@@ -143,4 +144,21 @@ public function scopeWithIsBeingWatched(Builder $builder, $user)
         });
     }
+
+    public function scopeWithSearch(Builder $builder, ?string $search)
+    {
+        $search = strtr($search, ['%' => '\%', '_' => '\_', '\\' => '\\\\']);
+
+        $builder->when(Features::enabled('search') && $search, function ($query) use ($search) {
+            $query->where(function ($query) use ($search) {
+                $query->where('title', 'like', '%'.$search.'%')
+                    ->orWhereExists(function ($query) use ($search) {
+                        $query->select(DB::raw(1))
+                            ->from('posts')
+                            ->whereColumn('discussion_id', 'discussions.id')
+                            ->where('content', 'like', '%'.$search.'%');
+                    });
+            });
+        });
+    }
```
@tomirons thoughts on this?

andreaselia commented on src/Models/Post.php:
```diff
@@ -49,4 +51,13 @@ public function getIsRichlyFormattedAttribute(): bool
         return $this->formatting === 'rich';
     }
+
+    public function scopeWithSearch(Builder $builder, ?string $search)
+    {
+        $search = strtr($search, ['%' => '\%', '_' => '\_', '\\' => '\\\\']);
+
+        $builder->when(Features::enabled('search') && $search, function ($query) use ($search) {
+            $query->where('content', 'like', '%'.$search.'%');
+        });
+    }
```
@tomirons and this?

andreaselia commented:
@Arkitecht would you mind adding some tests for this functionality please?

andreaselia commented:
You're awesome @Arkitecht, thanks for adding test coverage for this feature. Are you happy with the PR now? 🚀

Arkitecht commented:
I am good with it if you guys are! Have a couple more updates coming soon :)

andreaselia commented:
Just a few conflicts @Arkitecht from the previous PR, would you mind sorting them and then I can merge and tag a new release. I look forward to seeing the other updates coming soon, thank you!

Arkitecht commented:
I have addressed all of the conflicts, and added the search to the group page as well, with test coverage.

Let me know if you see anything else off

andreaselia commented:
Awesome, thanks @Arkitecht. It looks good to me so I'm going to merge it now, but won't tag a release until later as there are some general changes I may make to the package just before I tag the new release.

andreaselia commented:
I ended up publishing v1.6.0 now as I'm not sure when I'll get time to complete some frontend updates.

Arkitecht commented:
Awesome, thank you!

Teaser....
<img width="1296" alt="image" src="https://user-images.githubusercontent.com/66117/197367932-97507cb2-7400-4057-ac41-0271719cc676.png">

andreaselia commented:
> Awesome, thank you!
> 
> Teaser.... <img alt="image" width="1296" src="https://user-images.githubusercontent.com/66117/197367932-97507cb2-7400-4057-ac41-0271719cc676.png">

Reactions 😍 I look forward to seeing your PR for it 😁
